### PR TITLE
PLOT: Split matplotlib specific code from pandas plotting

### PR DESCRIPTION
**Pull request by datapythonista**
- [X] xref #14130
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

In order to be able to change pandas plotting backend, I think we need 3 main things:
1. Decouple the current matplotlib backend from the pandas plotting system, and just call it using a defined public API
2. Implement an option to select the backend (e.g. `pandas.set_option('plotting.backend', 'pandas.matplotlib')`)
3. Have other alternative backends which define the same API as the current matplotlib one

This PR is splitting most of the matplotlib backend code into a new `_matplotlib.py` file. Some matplotlib code still lives in `plotting/_core.py`, but would be nice to have early feedback on whether everybody is happy with this approach before moving forward. CC: @jreback @TomAugspurger

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Thanks, will take a look later. In case it's useful now, I started https://github.com/TomAugspurger/pandas/commit/b07aba28a37b0291fd96a1f571848a7be2b6de8d, which adds a config option for plotting, but never got to submitting a PR.

**Comment by jbrockmendel** [comment]
Totally on-board with the concept.  What other backends do you expect (either implemented here or downstream)?  My understanding is many of the other graphics libraries wrap matplotlib themselves.

**Comment by TomAugspurger** [comment]
I know of two libraries that have reimplemented pd.DataFrame.plot

- https://github.com/altair-viz/pdvega/
- https://hvplot.pyviz.org

**Comment by datapythonista** [comment]
I think Phillip started hvplot with the discussions of this issue, but as we didn't finish this work, it ended monkeypatching pandas.

The goal here is that projects like the ones Tom mentioned, and others like https://github.com/PatrikHlobil/Pandas-Bokeh can be implemented simpler and in a standard way, so we make the life of devs and users easier.

Personally, I'd at some point split the matplotlib backend from pandas and have it as a third party package. But that's for the long term, and unrelated to this PR.

**Comment by datapythonista** [comment]
For what I could see in the code, the API that pandas plotting backends need to provide are:

- `LinePlot`
- `BarPlot`
- `BarhPlot`
- `HistPlot`
- `BoxPlot`
- `KdePlot`
- `AreaPlot`
- `PiePlot` (only for `pandas.Series`)
- `ScatterPlot` (only for `pandas.DataFrame`)
- `HexBinPlot` (only for `pandas.DataFrame`)
- `hist_series`
- `hist_frame`
- `boxplot_frame`
- `boxplot_frame_groupby`

The implementation in this PR moves all the `matplotlib` code into `pandas.plotting.matplotlib`. And keeps all the public documentation in pandas (outside of the matplotlib backend). In a follow up PR should be simple to add an option to change the backend to another respecting the same API as the matplotlib one.

Any comment appreciated.

CC: @tacaswell, @jakevdp, @philippjfr, @PatrikHlobil

**Comment by jakevdp** [comment]
This is great - thanks for working on it!

File: pandas/plotting/matplotlib/__init__.py
Comment (TomAugspurger) [inline pandas/plotting/matplotlib/__init__.py]: Rename this module to `pandas/plotting/_matplotlib`

File: pandas/plotting/matplotlib/main.py
Comment (TomAugspurger) [inline pandas/plotting/matplotlib/main.py]: Maybe "core" instead of "main"? I typically associate main with `__main__` / running from the command line.

**Comment by tacaswell** [comment]
I am too :zzz: right now to have much sensible comments on the code.

If you are interested in pulling this whole thing out and putting it in `matplotlib/matplotlib-pandas` I am  open to that and would happily make sure anyone from the pandas side who needs / wants access will have it.  That is probably something to look at down the road after sorting out what you want the interface to be (mostly to avoid obligatory coordinated releases).

If there is interest in more aggressive re-thinking of the API https://github.com/DistrictDataLabs/yellowbrick/issues/826 , https://github.com/matplotlib/matplotlib/pull/14058, and https://paper.dropbox.com/doc/Matplotlib-4.0--AdQMb0ugRPWeqq3U4dEAHuSDAQ-WTYwd0NQaSHTjtLUZwkNx may be of interest.

**Review by jreback — CHANGES_REQUESTED** [review]
is this PR *just* moving things (and fixing the imports). would greatly appreciate that. Then can followon with a PR to change code. It is impossible to see what you are actually changing otherwise.

**Comment by datapythonista** [comment]
100% agree, there is no functional change, and no public API change in this PR. Just moving all the code specific to matplotlib from `pandas/plotting/*` to `pandas/plotting/_matplotlib/*`. All the calls to the "public" functions in `_matplotlib` are done by importing the module in a lazy way (that fails when matplotlib is not installed in the same was as it fails now).

Ideally would like to avoid touching any tests to guarantee that no unintentional functional change happens. In practice I need to update the import paths of some tests, since they import private functions that have been moved.

Still work in progress, the change is much bigger than what I originally thought, but will provide more details on what the PR involves when it's ready to be merged.

**Comment by jreback** [comment]
thanks @datapythonista yeah also happy to do this in pieces if its easier.

**Comment by datapythonista** [comment]
This should be ready now. This PR just moves code around, and shouldn't change the behavior of pandas at all. The implemented changes are:

- All the code that imports and uses matplotlib now lives in `pandas/plotting/_matplotlib`
- There is a clear API for what's in `pandas.plotting` and also in `pandas.plotting._matplotlib`, and no private methods are imported (except for tests, benchmarks and deprecated code)
- The changes in this PR should make very easy to change the matplotlib plotting backend by another one based in bokeh, holoviews, altair... as far as they are implemented using the same public API

Once this is merged I'll open issues to discuss the possible follow up changes:
- Using an option to specify the plotting backend
- Possible deprecations (there are up to 4 ways to plot the same thing in pandas, e.g. `df.plot(kind='box')`, `df.plot.box()`, `df.boxplot()` and `pandas.plotting.boxplot(df)`). Probably worth deprecating some, making the API for plotting backends simpler, and making users life easier by providing a single consistent way
- Possibly moving the matplotlib backend out of the pandas repo.
- Whether the backends should implement plots like `andrews_curves`, `autocorrelation_plot`,... or if they shouldn't be part of the pandas plotting API (but a third party).
- ...

File: pandas/plotting/_core.py
Comment (jreback) [inline pandas/plotting/_core.py]: this should be the first function in this module and/or put somewhere else

File: pandas/plotting/_core.py
Comment (jreback) [inline pandas/plotting/_core.py]: _get_plot_backend could be here (ignore my previous comment), to organize things. can you add an explanation block on what is the purpose of all this (also blank line separate before comments)

File: pandas/plotting/_matplotlib/__init__.py
Comment (jreback) [inline pandas/plotting/_matplotlib/__init__.py]: we tend to use absolute imports I think everywhere else

**Review by jreback — CHANGES_REQUESTED** [review]
looks pretty good; on followup (or just create issues), pls move the tests as much as possible to a generic testing framework (for any backend), and mpl specific tests.

**Comment by datapythonista** [comment]
Thanks for the review, addressed comments. Agree on splitting the tests, will follow up with that. I thought using relative imports made the matplotlib backend more decoupled (and easy to move if that's wanted at some point). Changed them to absolute imports for now, and we can see in the future if that's needed/useful.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. @jorisvandenbossche or @TomAugspurger

**Comment by TomAugspurger** [comment]
This seems to have lost our warning

```
In [1]: paste
import pandas as pd
import matplotlib.pyplot as plt
df = pd.Series(1, pd.date_range('2000', periods=12))

fig, ax = plt.subplots()

ax.plot(df.index, df.values)
## -- End pasted text --
/Users/taugspurger/sandbox/pandas/pandas/plotting/_converter.py:127: FutureWarning: Using an implicitly registered datetime converter for a matplotlib plotting method. The converter was registered by pandas on import. Future versions of pandas will require you to explicitly register matplotlib converters.

To register the converters:
        >>> from pandas.plotting import register_matplotlib_converters
        >>> register_matplotlib_converters()
  warnings.warn(msg, FutureWarning)
Out[1]: [<matplotlib.lines.Line2D at 0x10c614048>]
```

There's no warning on this branch.

I think that's because `import pandas` doesn't do an `import matplotlib` inside a try / except anymore. On master, `import pandas` would register our units / converters.

I don't recall when that deprecation is slated for removal. May as well be 0.24 though?

**Comment by TomAugspurger** [comment]
On the converter warning, that was added in 0.21. https://github.com/pandas-dev/pandas/pull/24964. Something happened, and it wasn't actually showing up till 0.24. So I'm not sure what's best here.

If you do want to enforce the deprecation, then I'd recommend merging this first, and cleaning up the deprecation in a followup.

**Comment by simonjayhawkins** [comment]
@datapythonista can you merge master.

Conflicting files
pandas/plotting/_compat.py

**Comment by datapythonista** [comment]
@TomAugspurger I restored the warning for now. The problem is that the `pandas.plotting._matplotlib` module is never loaded with your code, so registering the converters in that module, where it should happen is not possible.

I implemented it in a somehow hacky way. In `pandas.plotting` I test whether matplotlib is installed, and if it is, I register the converters (and load `pandas.plotting._matplotlib`). This is not idea, but we'll still be able to use other plotting backends without problems, and it's temporary (we can see later if we want to remove the automatic registration of converters in 0.25, 1.0 or when).

I think this solution is the best, let me know if you have a better idea (I'm not sure if this is what you were proposing).

**Comment by jreback** [comment]
can you merge master

**Comment by TomAugspurger** [comment]
>  I restored the warning for now.

Thanks. Looks good, and I'll do a followup issue to discuss removing it (I think after 0.25).

**Comment by PatrikHlobil** [comment]
> This should be ready now. This PR just moves code around, and shouldn't change the behavior of pandas at all. The implemented changes are:
> 
> * All the code that imports and uses matplotlib now lives in `pandas/plotting/_matplotlib`
> * There is a clear API for what's in `pandas.plotting` and also in `pandas.plotting._matplotlib`, and no private methods are imported (except for tests, benchmarks and deprecated code)
> * The changes in this PR should make very easy to change the matplotlib plotting backend by another one based in bokeh, holoviews, altair... as far as they are implemented using the same public API
> 
> Once this is merged I'll open issues to discuss the possible follow up changes:
> 
> * Using an option to specify the plotting backend
> * Possible deprecations (there are up to 4 ways to plot the same thing in pandas, e.g. `df.plot(kind='box')`, `df.plot.box()`, `df.boxplot()` and `pandas.plotting.boxplot(df)`). Probably worth deprecating some, making the API for plotting backends simpler, and making users life easier by providing a single consistent way
> * Possibly moving the matplotlib backend out of the pandas repo.
> * Whether the backends should implement plots like `andrews_curves`, `autocorrelation_plot`,... or if they shouldn't be part of the pandas plotting API (but a third party).
> * ...

Hi, thanks for working on this issue. If there is native support for a plotting backend switch, implementing new plotting solutions for Pandas will be much more attractive! 

Does every backend need to provide a this APIs or are they optional? This could, in my opinion) discourage people to implement their backend solution + restrict the creativity of the developers because each backend could provide some innovative ideas (eg matplotlib vs seaborn or bokeh vs holoviews).

Furthermore, is there also a minimal set of keyword arguments (x, y, color, xlabel, ...) that has to be implemented for each glyph such that some minimal examples work for each backend? Also, will the development of the plotting backends happen inside pandas-dev, a new pandas-plotting-dev or a own repository like hvplot/Pandas-Bokeh?

Best, Patrik

**Comment by TomAugspurger** [comment]
Backends should implement the same API, aside from the return type. For alternative APIs, libraries can accept DataFrames, just like seaborn.

**Comment by TomAugspurger** [comment]
@datapythonista are you able to reproduce the CI failures like https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=12231 locally? If not, I can try taking a look.

**Comment by datapythonista** [comment]
No, I'm taking a look, but very confused on what's going on, even the mypy error. The CI was green before merging master, which confuses me even more. If you can take a look too, that would be great, let me know if you find something.

**Comment by WillAyd** [comment]
@datapythonista haven't had the chance to look too deep but interesting that `boxplot_frame` throws a mypy error while `hist_frame` the line above doesn't. What's different between those two is the Appender - can you see if removing that maybe silences the mypy error?

**Comment by datapythonista** [comment]
thanks @WillAyd, I also saw that, and removing the Appender does fix the problem, but I think there is a root error, Appender is being used eveywhere... :\

**Comment by WillAyd** [comment]
Yea not sure why it's appearing now. Probably just need to annotate Appender for proper inference.

I don't think it's critical for this PR so if that's the only hangup feel free to add `type: ignore` and we can address in a follow up

**Comment by datapythonista** [comment]
I tried annotating `Appender` but didn't work. I saw a mypy issue where this error was raised for unrelated problems. One of the test failures (happens just with couple of builds) is related to boxplot too. I'll see if I can find the problem for that, as it's difficult to debug the mypy error. With some luck all the problems are caused by the same. Thanks for the help.

**Comment by WillAyd** [comment]
No problem. As mentioned I think OK to mark as type: ignore for now so it doesn't hold this up. I've opened #26629 to address separately

**Comment by datapythonista** [comment]
No idea what's going on here. The 3 errors seem unrelated. One of them is a failure on the qt backend of matplotlib because can't find the environment variable `DISPLAY` (I don't think this PR affects that variable, I guess the problem is something else causing that).

The other error is caused by a plot losing the tick information in a plot. None of them reproduce locally with the dependencies of the build (one is the mac build, may be someone with a mac can try to reproduce).

The mypy error seems to be nonsense too.

**Comment by TomAugspurger** [comment]
I tried on a Mac, and wasn't able to reproduce the error (will try in an exact replica of that env later). I will note that the warning shows up under pytest under your branch, whereas it didn't before. I'm not sure why that would be.

**Comment by datapythonista** [comment]
I fixed the errors, not sure why they were caused, but importing matplotlib to see if it was installed before registering the converters was causing the problems (including the mypy one which is weird).

CI is green, and it's generating the warning reported by @TomAugspurger like in master. Should be ready to be merged.

**Comment by jreback** [comment]
merge master one more time, ping on green.

File: pandas/plotting/_matplotlib/boxplot.py
Comment (jreback) [inline pandas/plotting/_matplotlib/boxplot.py]: later on I suppose could move all imports to the top as this is now a delayed import generally

**Comment by jreback** [comment]
@datapythonista after merge if you can create any needed followup issues, and issues for splitting tests and moving imports

**Comment by datapythonista** [comment]
@jreback this is green.

Will open PRs for the immediate follow up things, and create issues for the rest once merged.

**Comment by jreback** [comment]
thanks @datapythonista very nice!

**Pull request merged**
